src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdg_decoration_unstable_v1.h"
+
+#include "mir/log.h"
+#include "mir/shell/surface_specification.h"
+#include "mir/wayland/client.h"
+#include "mir/wayland/protocol_error.h"
+
+#include "xdg-decoration-unstable-v1_wrapper.h"
+#include "xdg_output_v1.h"
+#include "xdg_shell_stable.h"
+
+#include <memory>
+#include <unordered_set>
+
+namespace mir
+{
+namespace frontend
+{
+class ToplevelsWithDecorations
+{
+public:
+    ToplevelsWithDecorations() :
+        toplevels_with_decorations{std::unordered_set<wl_resource*>()}
+    {
+    }
```

Comment:
There's no need to copy-construct from a temporary. The only reason to write a default constructor would be disable copying:
```suggestion
    ToplevelsWithDecorations() = default;
    ToplevelsWithDecorations(ToplevelsWithDecorations const&) = delete;
    ToplevelsWithDecorations& operator=(ToplevelsWithDecorations const&) = delete;
```